src/modules/emotes_gamewisp/gw-emotes.js

```diff
@@ -0,0 +1,185 @@
+    }
+
+    getImageSizes(urlSmall) {
+        return {
+            '1x': urlSmall,
+            '2x': urlSmall.replace('28x28', '56x56'),
+            '4x': urlSmall.replace('28x28', '112x112')
+        };
+    }
+
+    initializeRoom(data) {
+        let emotes = data.emotes ? data.emotes : [];
+
+        emotes = emotes.map(emoteData => this.buildEmote(emoteData));
+
+        for (const username in data.userStore) {
+            if (data.userStore.hasOwnProperty(username)) {
+                let userEmotes = this.emotes.get(name);
+                if (!userEmotes) {
+                    userEmotes = new Map();
+                    this.emotes.set(username, userEmotes);
+                }
+
+                const emoteIDsSet = new Set(data.userStore[username]);
+
+                emotes.forEach(e => {
+                    if (emoteIDsSet.has(e.id)) {
+                        userEmotes.set(e.code, e);
+                    }
+                });
+            }
+        }
+    }
+
+    updateRoom(data) {
+        let newEmotes = data.emotes;
+        const newUser = data.user;
+
+        newEmotes = newEmotes.map(emoteData => this.buildEmote(emoteData));
+
+        let userEmotes = this.emotes.get(newUser.name);
+        if (!userEmotes) {
+            userEmotes = new Map();
+            this.emotes.set(newUser.name, userEmotes);
+        }
+
+        const userEmoteIDsSet = new Set(newUser.emoteIDs);
+
+        newEmotes.forEach(e => {
+            if (userEmoteIDsSet.has(e.id)) {
+                userEmotes.set(e.code, e);
+            }
+        });
+    }
+
+    leaveRoom(data) {
+        const username = data.user;
+
+        if (this.emotes.get(username)) {
+            this.emotes.delete(username);
+        }
+    }
+
+    removeEmote(data) {
+        const code = data.emoteCode;
+        const emotes = [...this.emotes.values()];
+
+        emotes.forEach(emoteMap => {
+            emoteMap.delete(code);
+        });
+    }
+
+    addEmote(data) {
+        const emote = this.buildEmote(data.emote);
+
+        let userEmotes;
+
+        data.emoteUsers.forEach(user => {
+            if (this.emotes.has(user)) {
+                userEmotes = this.emotes.get(user);
+            } else {
+                userEmotes = new Map();
+                this.emotes.set(user, userEmotes);
+            }
+
+            userEmotes.set(emote.code, emote);
+        });
+    }
+
+    newSubscriber(data) {
+        const newSub = data.user;
+        let userEmotes;
+
+        if (this.emotes.has(newSub)) {
+            userEmotes = this.emotes.get(newSub);
+        } else {
+            userEmotes = new Map();
+            this.emotes.set(newSub, userEmotes);
+        }
+
+        data.emotes.forEach(gwEmote => {
+            userEmotes.set(gwEmote.code, this.buildEmote(gwEmote));
+        });
+    }
+
+    cancelSubscriber(data) {
+        if (!this.emotes.has(data.user)) return;
+
+        const emoteIDsToRemove = new Set(data.emoteIDs);
```

Comment:
unrelated, but from an API perspective this is kinda lame.. you could just emit an "update_user" to handle the sub/unsub case by returning a list of the users' current emote ids